torch/csrc/tensor/python_tensor.cpp

```diff
@@ -405,7 +405,11 @@ void py_set_default_dtype(PyObject* obj) {
       [backend, scalar_type](PyTensorType *x) {
         return x->get_backend() == backend && x->get_scalar_type() == scalar_type;
       });
-    set_default_tensor_type(*it);
+    if (it != tensor_types.end()) {
+      set_default_tensor_type(*it);
+    } else {
+      throw TypeError("tensor type not found");
```

Comment:
we should throw the error with the same message as for other non-floating point types
```
>>> torch.set_default_dtype(torch.int16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/fsx/users/chourdiaanjali/tree2/torch/__init__.py", line 369, in set_default_dtype
    _C._set_default_dtype(d)
TypeError: only floating-point types are supported as the default type
```

Reply:
I think the real issue is that the `scalar_types` in this function
https://github.com/pytorch/pytorch/blob/224093db113fd333a7cb8e5e07308de5098e4a62/torch/csrc/utils/tensor_types.cpp#L84 doesn't include complex types (and it should). Currently any dtype not in that list also leads to a segfault, for example:
```
>>> torch.set_default_dtype(torch.qint32)
Segmentation fault (core dumped)
```
. We should consider keeping this list in sync with `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS` https://github.com/pytorch/pytorch/blob/224093db113fd333a7cb8e5e07308de5098e4a62/c10/core/ScalarType.h#L29